packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,315 @@
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function buyTokens(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures)
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        require(msg.value > 0, ERROR_INVALID_INPUT);
+        address token = readAddress(orders[0].takerAssetData, 1);
+        require(token == address(ETHER_TOKEN), ERROR_INVALID_INPUT);
+
+        ETHER_TOKEN.deposit.value(msg.value)();
+        Exchange.FillResults memory fillTokensFillResult = marketSellTokens(orders, signatures, feeOrders, feeSignatures, msg.value);
+        addFillResults(totalFillResult, fillTokensFillResult);
+        return totalFillResult;
+    }
+
+    /// @dev Buys the tokens, performing fee abstraction if required. This function is payable
+    ///      and will convert all incoming ETH into WETH and perform the trade on behalf of the caller.
+    ///      This function allows for a deduction of a proportion of incoming ETH sent to the feeRecipient.
+    ///      The caller is sent all tokens from the marketSell of orders (less any fees required to be paid in ZRX)
+    ///      If the purchased token amount does not meet some threshold (98%) then this function reverts.
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeProportion A proportion deducted off the incoming ETH and sent to feeRecipient. The maximum value for this
+    ///        is 1000, aka 10%. Supports up to 2 decimal places. I.e 0.59% is 59.
+    /// @param feeRecipient An address of the fee recipient whom receives feeProportion of ETH.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function buyTokensFee(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint16  feeProportion,
+        address feeRecipient)
+        payable
+        public
+        returns (Exchange.FillResults memory totalFillResult)
+    {
+        require(msg.value > 0, ERROR_INVALID_INPUT);
+        address token = readAddress(orders[0].takerAssetData, 1);
```

Comment:
We chatted previously about having a decode library. I created [a task](https://app.asana.com/0/636508596987970/651649276736967) for myself so this doesn't get lost.